### PR TITLE
[exporter/coralogix] Create dynamic subsystem name

### PR DESCRIPTION
**Pull request by ofirshmuel**
**Description:** <Describe what has changed.>

Create dynamic subsystem name for Coralogix endpoint.
The subsystem name is taken from the service_name of span.
Add exporter timeout option.

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Review by oded-dd — APPROVED** [review]

File: exporter/coralogixexporter/client.go [hunk 1]
```diff
@@ -70,13 +70,20 @@ func (c *coralogixClient) newPost(ctx context.Context, td pdata.Traces) error {
...
-	ctx = metadata.NewOutgoingContext(ctx, metadata.New(c.cfg.GRPCClientSettings.Headers))
-	for _, batch := range batches {
 
+	for _, batch := range batches {
+		subsystem := batch.GetProcess().ServiceName
+		c.metadata = cxpb.Metadata{
+			ApplicationName: c.cfg.AppName,
+			SubsystemName:   subsystem,
+		}
+		c.cfg.GRPCClientSettings.Headers["subsystemName"] = subsystem
```
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:80]: If your user has set this header themselves in the config, you are overriding it without warning. Be nice to your users and inform them that this is happening.

File: exporter/coralogixexporter/config.go [hunk 2]
```diff
@@ -39,9 +39,8 @@ type Config struct {
 	PrivateKey string `mapstructure:"private_key"`
 
 	// Traces emitted by this OpenTelemetry exporter should be tagged
-	// in Coralogix with the following application and subsystem names
-	AppName   string `mapstructure:"application_name"`
-	SubSystem string `mapstructure:"subsystem_name"`
```
Comment (jpkrohling) [inline exporter/coralogixexporter/config.go:44]: This is effectively a breaking change and has to be done following these rules: https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#breaking-changes

File: exporter/coralogixexporter/client.go (hunk 1 above)
Comment (ofirshmuel) [reply exporter/coralogixexporter/client.go:83]: A warning message is added

**Comment by oded-dd** [comment]
Hi @jpkrohling any update on the PR?

**Comment by ofirshmuel** [comment]
searching for an issue with problems, freezing for now

**Comment by ofirshmuel** [comment]
Hi @jpkrohling any updates?

**Comment by jpkrohling** [comment]
See my comments about the config breaking compatibility with the previous version.

**Review by oded-dd — CHANGES_REQUESTED** [review]

**Review by oded-dd — APPROVED** [review]

**Comment by ofirshmuel** [comment]
Hi @jpkrohling, according to your comment about breaking changes. I add the changes to the changelog file. In addition, I update the readme, tests and config_test file. The 'subsystems' config option is not available anymore.

**Comment by jpkrohling** [comment]
@ofirshmuel, please go over the document I shared earlier: https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#breaking-changes

The document above has concrete steps you should be taking to prevent breaking your users without notice. As it is, just removing the config option isn't acceptable without deprecating it first.

**Comment by ofirshmuel** [comment]
@jpkrohling I mark the `subsystem` as deprecated and add more comments about it.
In the next version, I will open PR to remove it as mentioned in contributing document.

File: exporter/coralogixexporter/client.go [hunk 3]
```diff
@@ -70,12 +65,20 @@ func (c *coralogixClient) newPost(ctx context.Context, td pdata.Traces) error {
...
-		_, err := c.client.PostSpans(ctx, &cxpb.PostSpansRequest{Batch: *batch, Metadata: &c.metadata}, grpc.WaitForReady(false))
-		if err != nil {
-			return fmt.Errorf("failed to push trace data via Coralogix exporter %w", err)
+		if batch.GetProcess().GetServiceName() != "" {
+			_, err := c.client.PostSpans(ctx, &cxpb.PostSpansRequest{
+				Batch:    *batch,
+				Metadata: &cxpb.Metadata{ApplicationName: c.cfg.AppName, SubsystemName: batch.GetProcess().GetServiceName()},
+			}, grpc.WaitForReady(c.cfg.WaitForReady))
+			if err != nil {
+				return fmt.Errorf("Failed to push trace data via Coralogix exporter %w", err)
```
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:77]: Please, follow the Go conventions by not uppercasing the first letter from the error message

File: exporter/coralogixexporter/client.go [hunk 4]
```diff
@@ -70,12 +65,20 @@ func (c *coralogixClient) newPost(ctx context.Context, td pdata.Traces) error {
...
+			}
+			c.logger.Debug("Trace was sent successfully")
+		} else {
+			return fmt.Errorf("Failed to push trace data via Coralogix exporter because batch.process.serviceName is empty")
```
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:81]: Same here and possibly elsewhere.

File: exporter/coralogixexporter/config.go [hunk 5]
```diff
@@ -56,7 +58,7 @@ func (c *Config) Validate() error {
 		return fmt.Errorf("`appName` not specified, please fix the configuration file")
 	}
 	if c.SubSystem == "" {
-		return fmt.Errorf("`subSystem` not specified, please fix the configuration file")
+		return fmt.Errorf("`subSystem` not specified, please fix the configuration file.\nPay attention: deprecated subsystem will be removed in the next version")
```
Comment (jpkrohling) [inline exporter/coralogixexporter/config.go:61]: Imagine error messages in a broader context: they bubble up and are printed to the user by the component that is the closest to the user, collecting all the breadcrumbs until it gets there. In that context, new lines aren't desirable, as your message would end up looking like this:

    2022-03-11 14:10:12.1234 ERR could not start blurbs: failed to parse config file: `subSystem` not specified, please fix the configuration file
    Pay attention: deprecated subsystem will be removed in the next version.

That said, you should really make room for your users to migrate to the replacement already. If they won't be allowed to specify this option, then make it possible for your users to not specify this already and leave their config files ready for the next version.

File: exporter/coralogixexporter/config.go [hunk 6]
```diff
@@ -40,7 +41,8 @@ type Config struct {
 
 	// Traces emitted by this OpenTelemetry exporter should be tagged
 	// in Coralogix with the following application and subsystem names
-	AppName   string `mapstructure:"application_name"`
+	AppName string `mapstructure:"application_name"`
+	// Deprecated: [v0.47.0] SubSystem will remove in the next version
```
Comment (jpkrohling) [inline exporter/coralogixexporter/config.go:45]: Specify also what users should be doing instead. Is it safe to just remove this from the config file? Or will they need to do something else?

Comment (ofirshmuel) [reply exporter/coralogixexporter/config.go:45]: application_name should stay on the config file, subsystem_name  will be removed.
The users need to remove 'subsystem_name' in the next version, the 'subsystem_name' will generate automatically according to the "service_name" of the batch. I will add a description to the comments.

File: exporter/coralogixexporter/client.go (hunk 3 above)
Comment (ofirshmuel) [reply exporter/coralogixexporter/client.go]: I changed the uppercase to lowercase according to the conventions.

File: exporter/coralogixexporter/client.go (hunk 4 above)
Comment (ofirshmuel) [reply exporter/coralogixexporter/client.go]: I changed the uppercase to lowercase according to the conventions.

File: exporter/coralogixexporter/config.go (hunk 5 above)
Comment (ofirshmuel) [reply exporter/coralogixexporter/config.go]: thanks @jpkrohling , I fixed the error message and let the users add or remove 'subsystem' option from the config file in this version.

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**